### PR TITLE
Proj02 - completed gamemodes, pause, and leaderboards

### PR DESCRIPTION
**Pull request by andyyy60**

### CODE REVIEW

File: src/edu/ucsb/cs56/projects/games/subsink/Screen.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    menuBounds.put(text, new Rectangle(x, y, width + 10, height + 10));
+                    y += height + 10 + 5;
+                }
+
+            }
+            for (String text : menuItems) {
+                Rectangle bounds = menuBounds.get(text);
+                boolean isSelected = text.equals(selectedMenuItem);
+                boolean isFocused = text.equals(focusedItem);
+                if (selectedMenuItem.equals("Start Game") || selectedMenuItem.equals("Try Again")) {
```
Comment (rhelmot) [inline src/edu/ucsb/cs56/projects/games/subsink/Screen.java]: This behavior-based code should _absolutely not_ be part of the paint routine. It should be part of the callback that actually handles the click.

**Comment by rhelmot** [comment]
Yo, so there are some serious issues.

1. You need to give attribution for the huge class you copied from stack overflow. this is important for passing the code onto future generations, and is also important from a let's-not-plagiarize-things perspective. It's also a licensing issue - code from stack overflow is licensed as [cc-by-sa](https://creativecommons.org/licenses/by-sa/3.0/), which means you have to give attribution and also license your work under the same license. The license-your-own-work-the-same thing isn't a huge issue for a school project like this, but it's something to know about.
2. You seem to not really have a design in mind for organizing the modal screens. Some of them are JFrames and some are JPanels. Pick one or the other. I'd prefer the latter.
3. The modal screens just don't work currently. I click the leaderboard and how to play buttons and there's a 50/50 chance that either the screen or the back button show up, always exactly one, never both. This just means you're misusing the swing APIs.
4. You have this design pattern again where just creating a new object causes its state to be registered somewhere - this is not a good thing. Object creation should be free from side effects. A good example of this is how I split the original classes into the constructor and then the run() method. Programmers generally understand that `new Class()` ought to just create a new object, but `method.doSomething()` might affect state in far-reaching places.

These are all things I'd take points off from. As promised during lab hours, I'll give you one extra day past the deadline for submission, since I took a full day to get back to this. You can come to my lab hours on thursday for help, and I can also help you with this stuff during section. You can also just ask questions here and I'll get to them when I can.

**Comment by andyyy60** [comment]
Hi Audrey,

I have addressed those issues in the newest push to branch: proj02

1. I completely scraped such class and made my own, it looks much uglier
but it's mine now. This caused me to create an abstract Screen class and a
GameOverScreen, StartScreen which implement Screen. Leaderboard and High
Scores stayed the same.
2. Picked JFrames for all Screen related stuff and JButtons for buttons
3. I believe this is fixed with the new screen implementation? Couldn't
really replicate the bug
4. Every class now has a .run() method where the user explicitly has to
tell the class to be run. Nothing happens now when an object is simply
created. This required a change in all the screens + in World and Main
where I called such classes

One last thing, I accidentally pushed this to proj01 branch, but I reverted
the changes now using the trick you taught me last time (reset HEAD and
force push), now I *believe* its fixed. Sorry, its late and I'm tired.

I really hope this satisfies the points now (for proj02), I will be working
on the smart subs problem tomorrow to make up points if cs170 allows me to.

Andy.

**Comment by rhelmot** [comment]
Hey durva - do you have unpushed changes for the intelligent subs stuff? the stuff present right now a) contains bugs which make it identical to the old behavior, and b) crashes whenever you drop a depth charge.

**Comment by rhelmot** [comment]
Hey andy - the extra screens are not working for me. When I click "high scores" or "how to play", I _only_ see a back button. On the high scores screen, the back button closes the window, but on the how to play screen, the back button replaces the contents of the window with another menu screen. Finally, when I finish the game, the score entry pops up in a new window, and the game is left running without the ship on the old screen.

image

- top left: original screen, screen I played the game in
- top right: high scores screeen
- bottom right: how to play screen after pressing the back button. before pressing the back button it looked identical to the screen above it.
- bottom left: screen generated after beating the game

Are any of these different from how it works on your computer? We may be running into platform-specific issues.

**Comment by andyyy60** [comment]
Hi Audrey,

That's very odd. Here are some screenshots from my machine:



I'm not sure too sure about the smart subs though.

Andy.

**Comment by rhelmot** [comment]
I don’t think I out can an post images in issue comments via email so they’re not showing up for me, can you post them directly to Github?

**Comment by rhelmot** [comment]
I don’t think I out can an post images in issue comments via email so they’re not showing up for me, can you post them directly to Github?

**Comment by andyyy60** [comment]
Here's the image:
image

**Comment by rhelmot** [comment]
So I talked to the professor about this.

The bottom line is that the system by which we do our grading around here is CSIL, and CSIL can't run the code you've submitted. What we can do is merge this code into a w18 branch on the uscb-cs56-projects account and keep it around as a reference for the next group to work on this, since it is good code and I'd like you to be rewarded for that, but it can't be merged to master. I can even do some git-fu to parse out the commits for the issues that you did correctly solve, maintaining authorship information and everything, so you should get full credit for those.

Points will be delivered later today.

**Comment by andyyy60** [comment]
Thanks so much Audrey. That's a bummer that the changes won't get reflected
in the master branch. I had no idea that my changes were platform
specific.. maybe that should be something to add to the repo so the next
group of students don't do the same mistake.

Thank you for being fair, I really did put a lot of time into this project,
and it makes me feel a little relaxed to know that points will still be
awarded.

Thanks,
Andy

**Pull request closed without merging**